internal/pkg/jwt/jwt.go

```diff
@@ -0,0 +1,240 @@
+	if requireBearerToken {
+		if publicKeyEndpoint == "" {
+			return nil, errors.New("missing public key endpoint for jwt")
+		}
+		if verifier == nil {
+			return nil, errors.New("missing jwt verifier")
+		}
+		if domain == "" || component == "" {
+			return nil, errors.New("missing jwt domain or component")
+		}
+	}
+	sc := DefaultJWTConsumer{
+		publicKeyEndpoint:  publicKeyEndpoint,
+		verifier:           verifier,
+		requireBearerToken: requireBearerToken,
+		keys:               make(map[string]*rsa.PublicKey),
+		domain:             domain,
+		component:          component,
+		adminClientIds:     adminClientIds,
+		capabilityPrefixes: capabilityPrefixes,
+		tenantStorer:       tenantStorer,
+		client: &http.Client{
+			Timeout: 30 * time.Second,
+		},
+	}
+	return &sc, nil
+}
+
+// getKeyData checks cached token and downloads keys from key vault if expired
+func (sc *DefaultJWTConsumer) getPublicKey(kid string) (*rsa.PublicKey, error) {
+	sc.Lock()
+	defer sc.Unlock()
+	pub, has := sc.keys[kid]
+	if has {
+		return pub, nil
+	}
+	// fetch from key vault
+	req, _ := http.NewRequest("GET", sc.publicKeyEndpoint+"/"+kid, nil)
+	resp, err := sc.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	bs, err := io.ReadAll(resp.Body)
+	if nil != err {
+		if resp.StatusCode == 404 {
+			return nil, &UnauthorizedError{InvalidKid}
+		} else {
+			return nil, err
+		}
+	}
+	if 200 != resp.StatusCode {
+		//ctx.Logger().Error("op", "JWTConsumer.getKeyData", "stage", "client.DoRequest", "status", status)
+		return nil, fmt.Errorf("Invalid status code: %d\n", resp.StatusCode)
+	}
+	pub, err = jwt.ParseRSAPublicKeyFromPEM(bs)
+	if nil != err {
+		//ctx.Logger().Error("op", "JWTConsumer.getKeyData", "stage", "decode", "err", err)
+		return nil, err
+	}
+	sc.keys[kid] = pub
+	return pub, nil
+}
+
+// VerifyToken validates a give token for an optional list of clientIds / "subjects" (use empty array if any client id ok) and
+// for a given domain, component, api and method (api and method are encoded in the claim)
+func (sc *DefaultJWTConsumer) VerifyToken(ctx context.Context, token string, api string, method string, tid *tenant.Id) ([]string, string, error) {
+	if !sc.requireBearerToken {
+		return nil, "", nil
+	}
+	if token == "" {
+		return nil, "", &UnauthorizedError{MissingToken}
+	}
+	partners, sub, capabilities, err := sc.extractToken(token)
+	if nil != err {
+		return nil, "", err
+	}
+	// verify clientId and partner
+	if "" == sub {
+		return nil, "", &UnauthorizedError{MissingClientId}
+	}
+	foundClientId := false
+	// check if this is an admin client
+	if 0 < len(sc.adminClientIds) {
+		for _, clientId := range sc.adminClientIds {
+			if sub == clientId {
+				foundClientId = true
+				break
+			}
+		}
+	}
+	// otherwise check if this is an app client
+	if !foundClientId {
+		tenantConfig, err := sc.tenantStorer.GetConfig(ctx, *tid)
```

Comment:
Now we have two places (one in ears and one in our flows api db) where we store tenant specific client ids. This doesn't seem optimal. Is there a better way to do this?

Reply:
yes that's right - the main problem here is that we want to keep ears clean of any internal dependencies so I'm not sure how else we can do this...